Reflect latest changes in synthetic source documentation (#109501)
elastic/elasticsearch: docs/reference/mapping/fields/synthetic-source.asciidoc

While this on the fly reconstruction is *generally* slower than saving the source
documents verbatim and loading them at query time, it saves a lot of storage
space. Additional latency can be avoided by not loading `_source` field in queries when it is not needed.

[[synthetic-source-fields]]
===== Supported fields
Synthetic `_source` is supported by all field types. Depending on implementation details, field types have different properties when used with synthetic `_source`.

> <<synthetic-source-fields-native-list, Majority of field types>> natively support synthetic `_source`. Source for the field is constructed using existing data, most commonly <<doc-values,`doc_values`>>, meaning no additional space is used.

Reviewer: Nit: s/Majority/ The majority/ ?

Reviewer 2: Maybe "Source for the field is usually reconstructed using doc values. However, sometimes stored fields are used, which can result in slight storage overhead."

> <<synthetic-source-fields-native-list, Majority of field types>> natively support synthetic `_source`. Source for the field is constructed using existing data, most commonly <<doc-values,`doc_values`>>, meaning no additional space is used.

Reviewer 3: ```suggestion
<<synthetic-source-fields-native-list,Most field types>> natively support synthetic `_source`. The source for natively supported fields is constructed using existing data, most commonly <<doc-values,`doc_values`>>, meaning that no additional space is used.
```

> If a field type does not have native synthetic `_source` support, original data of this field is stored as is in the same way as the `_source` field in non-synthetic mode. Similarly, malformed values of fields that use <<ignore-malformed,`ignore_malformed`>> need to be stored as is, requiring additional storage space.

Reviewer 4: Maybe add that the part of the _source is then stored in a special source field called `_ignore_source`?

Reviewer 2: I think we either explain somewhere the difference between native and non-native synthetic source...so that we can use the terminology later or we don't use such terminology and explain it differently like for instance:

"If a field type supports document reconstruction by means of doc values...otherwise, if doc values are not available the field value is stored as is..."

I am not in favor of one or the other...just saying that if we use "native synthetic source" we should explain it beforehand...or don't use it at all.

<<synthetic-source-fields-native-list, Most field types>> construct synthetic `_source` using existing data, most commonly <<doc-values,`doc_values`>> and <<stored-fields, stored fields>>. As such no additional space is used to store contents of `_source` field. Due to the storage layout of <<doc-values,`doc_values`>>, generated `_source` field undergoes <<synthetic-source-modifications, modifications>> compared to original document.

For all other fields, the original value of the field is stored as is, in the same way as the `_source` field in non-synthetic mode. In this case there are no modifications and field data in `_source` is the same as in the original document. Similarly, malformed values of fields that use <<ignore-malformed,`ignore_malformed`>> or <<ignore-above,`ignore_above`>> need to be stored as is. This approach is less storage efficient since data needed for `_source` reconstruction is stored in addition to other data required to index the field (like `doc_values`).

[[synthetic-source-restrictions]]
===== Synthetic `_source` restrictions

> There are some restrictions with usage of synthetic `_source`. Synthetic `_source` can not be used together with field mappings that use <<copy-to,`copy_to`>>. There are additional restrictions specific to field types. They are documented in synthetic `_source` section of the respective field type documentation page.

Reviewer 5: ```suggestion
There are some restrictions with usage of synthetic `_source`. Synthetic `_source` cannot be used together with field mappings that use <<copy-to,`copy_to`>>. There are additional restrictions specific to field types. They are documented in the synthetic `_source` section of the respective field type documentation page.
```

> There are some restrictions with usage of synthetic `_source`. Synthetic `_source` cannot be used together with field mappings that use <<copy-to,`copy_to`>>. There are additional restrictions specific to field types. They are documented in the synthetic `_source` section of the respective field type documentation page.

Reviewer 3: let's show them where to find the docs

```suggestion
Synthetic `_source` cannot be used together with field mappings that use <<copy-to,`copy_to`>>. 

Some field types have additional restrictions. These restrictions are documented in the **synthetic `_source`** section of the field type's <<mapping-types,documentation>>.
```

While this on the fly reconstruction is *generally* slower than saving the source
documents verbatim and loading them at query time, it saves a lot of storage
space. Additional latency can be avoided by not loading `_source` field in queries when it is not needed.

[[synthetic-source-fields]]
===== Supported fields
Synthetic `_source` is supported by all field types. Depending on implementation details, field types have different properties when used with synthetic `_source`.

> <<synthetic-source-fields-native-list, Most field types>> construct synthetic `_source` using existing data, most commonly <<doc-values,`doc_values`>> and <<stored-fields, stored fields>>, meaning no additional space is used to store contents of `_source` field. Generated `_source` field in this case undergoes <<synthetic-source-modifications, modifications>> compared to original document caused by the storage layout of <<doc-values,`doc_values`>>.

Reviewer: Nit: these lines are quite long, mind breaking them to shorter ones?

> For all other fields, original value of the field is stored as is in the same way as the `_source` field in non-synthetic mode. In this case there are no modifications and field data in `_source` is the same as in original document. Similarly, malformed values of fields that use <<ignore-malformed,`ignore_malformed`>> or <<ignore-above,`ignore_above`>> need to be stored as is. This approach is less storage efficient since data needed for `_source` reconstruction is stored in addition to other data required to index the field (like `doc_values`).

Reviewer: Nit:
s/original value of the field/the original value of a field/
s/as is in the same way/as is, in the same way/

Author: Appreciate these!

[[synthetic-source-restrictions]]
===== Synthetic `_source` restrictions

There are some restrictions with usage of synthetic `_source`:

> - Synthetic `_source` can not be used together with field mappings that use <<copy-to,`copy_to`>>.

Reviewer: Nit: since there's a single bullet, consider merging with the sentence above.

** <<numeric-synthetic-source,`long`>>
** <<range-synthetic-source,`range` types>>
** <<numeric-synthetic-source,`scaled_float`>>
** <<numeric-synthetic-source,`short`>>

> ** <<text-synthetic-source,`text`>>

Author: `text` is controversial - if there is no keyword subfield it will store the text

Reviewer: We should have a category "It's complicated" :P

Reviewer 4: imo native means the field type handling synthetic source without relying on ignored source. It says nothing about how it implements synthetic source.

[[synthetic-source-modifications-ranges]]
====== Representation of ranges

> Range field vales (e.g. `long_range`) are always represented as inclusive on both sides with bounds adjusted accordingly. See <<range-synthetic-source-inclusive, examples>>.

Reviewer 3: should this say `values`?

[[synthetic-source-precision-loss-for-point-types]]
====== Reduced precision of `geo_point` values
Values of `geo_point` fields are represented in synthetic `_source` with reduced precision. See <<geo-point-synthetic-source, examples>>.

[[synthetic-source-fields-native-list]]
===== Field types that support synthetic source with no storage overhead

> The following field types support synthetic source using data from <<doc-values,`doc_values`>> or <<stored-fields, stored fields>>, and require no additional storage space to construct the `_source` field. Additional storage is required to store ignored field values produced as a result of using <<ignore-malformed,`ignore_malformed`>> or <<ignore-above,`ignore_above`>> parameters.

Reviewer 3: I'd split this info into a note so it's clearer that it's supplemental

```suggestion
The following field types support synthetic source using data from <<doc-values,`doc_values`>> or <<stored-fields, stored fields>>, and require no additional storage space to construct the `_source` field. 

NOTE: If you enable the <<ignore-malformed,`ignore_malformed`>> or <<ignore-above,`ignore_above`>> settings, then additional storage is required to store ignored field values for these types.
```

[[synthetic-source-modifications-ranges]]
====== Representation of ranges
Range field vales (e.g. `long_range`) are always represented as inclusive on both sides with bounds adjusted accordingly. See <<range-synthetic-source-inclusive, examples>>.

[[synthetic-source-precision-loss-for-point-types]]
====== Reduced precision of `geo_point` values
Values of `geo_point` fields are represented in synthetic `_source` with reduced precision. See <<geo-point-synthetic-source, examples>>.

[[synthetic-source-fields-native-list]]

> ===== Field types that support synthetic source with no storage overhead

Reviewer: I'm on the fence on whether to report the fields that have additional indexing overhead when `ignore_malformed` or `ignore_above` are set. Wdyt?

Author: I think it's fair, i imagine normally you expect malformed values to be a small portion of your data.

Reviewer: Well, the copying parser always copies the current value preemptively, iiuc?

Author: I see what you mean. I don't know if it should be in this section but we can mention it somewhere. Let me follow up.

Reviewer 4: Let's wait with mentioning the copy parser overhead until we have a better understanding how much it actually is? I suspect it is minimal.

> Following field types support synthetic source using data from <<doc-values,`doc_values`>> and <<stored-fields, stored fields>> and require no additional storage space to construct `_source` field. Additional storage is required to store ignored field values produced as a result of using <<ignore-malformed,`ignore_malformed`>> or <<ignore-above,`ignore_above`>> parameters.

Reviewer: s/Following/The following/
